core/src/main/java/io/questdb/griffin/engine/table/TableReaderTimeFrameCursor.java

```diff
@@ -0,0 +1,232 @@
+        if (++partitionIndex < partitionHi) {
+            long timestampLo = reader.getPartitionTimestampByIndex(partitionIndex);
+            timeFrame.of(partitionIndex, timestampLo, estimatePartitionHi(timestampLo));
+            return true;
+        }
+        // Update frame index in case of subsequent prev() call.
+        timeFrame.of(partitionHi, Long.MIN_VALUE, Long.MIN_VALUE);
+        return false;
+    }
+
+    public TimeFrameRecordCursor of(DataFrameCursor dataFrameCursor) {
+        this.dataFrameCursor = dataFrameCursor;
+        reader = dataFrameCursor.getTableReader();
+        recordA.of(reader);
+        recordB.of(reader);
+        partitionHi = reader.getPartitionCount();
+        partitionCeilMethod = PartitionBy.getPartitionCeilMethod(reader.getPartitionedBy());
+        timestampIndex = reader.getMetadata().getTimestampIndex();
+        toTop();
+        return this;
+    }
+
+    @Override
+    public long open() throws DataUnavailableException {
+        int partitionIndex = timeFrame.partitionIndex;
+        if (partitionIndex < 0 || partitionIndex > partitionHi - 1) {
+            throw CairoException.nonCritical().put("open call on uninitialized time frame");
+        }
+        long size = reader.openPartition(timeFrame.partitionIndex);
+        if (size > 0) {
+            timeFrame.rowLo = 0;
+            timeFrame.rowHi = size;
+            final MemoryR column = reader.getColumn(TableReader.getPrimaryColumnIndex(reader.getColumnBase(timeFrame.partitionIndex), timestampIndex));
+            timeFrame.timestampLo = column.getLong(0);
+            timeFrame.timestampHi = column.getLong((size - 1) * 8) + 1;
+            return size;
+        }
+        timeFrame.rowLo = 0;
+        timeFrame.rowHi = 0;
+        timeFrame.timestampLo = timeFrame.estimateTimestampLo;
+        timeFrame.timestampHi = timeFrame.estimateTimestampLo;
+        return 0;
+    }
+
+    @Override
+    public boolean prev() {
+        int partitionIndex = timeFrame.partitionIndex;
+        if (--partitionIndex >= 0) {
+            long timestampLo = reader.getPartitionTimestampByIndex(partitionIndex);
+            timeFrame.of(partitionIndex, timestampLo, estimatePartitionHi(timestampLo));
+            return true;
+        }
+        // Update frame index in case of subsequent next() call.
+        timeFrame.of(-1, Long.MIN_VALUE, Long.MIN_VALUE);
+        return false;
+    }
+
+    @Override
+    public void recordAt(Record record, long atRowId) {
+        ((TableReaderSelectedColumnRecord) record).jumpTo(Rows.toPartitionIndex(atRowId), Rows.toLocalRowID(atRowId));
+    }
+
+    @Override
+    public void toTop() {
+        timeFrame.clear();
+    }
+
+    private long estimatePartitionHi(long partitionTimestamp) {
+        // partitionCeilMethod is null in case of partition by NONE
+        return partitionCeilMethod != null ? partitionCeilMethod.ceil(partitionTimestamp) : Long.MAX_VALUE;
```

Comment:
This is quite inaccurate in the case of split partitions. PartitionHi is the timestamp of the next partition.

Reply:
> PartitionHi is the timestamp of the next partition.

This is true only in the case of split partitions. What's the best way to detect this case?